File: step_16_dynamic_programming/16.6_dp_on_stocks/P001_best_time_to_buy_and_sell_i/solution.py
"""
Problem: Best Time to Buy and Sell Stock I (LeetCode #121)
Difficulty: EASY | XP: 10

Given array prices, find max profit from one buy and one sell.
"""
from typing import List


# ============================================================
# APPROACH 1: BRUTE FORCE (All Pairs)
# Time: O(n^2) | Space: O(1)
# ============================================================
def max_profit_brute(prices: List[int]) -> int:
    """Check every (buy, sell) pair where buy < sell."""
    max_profit = 0
    n = len(prices)

    for i in range(n - 1):
        for j in range(i + 1, n):
            profit = prices[j] - prices[i]
            max_profit = max(max_profit, profit)

    return max_profit


# ============================================================
# APPROACH 2: OPTIMAL (Single Pass with Running Minimum)
# Time: O(n) | Space: O(1)
# ============================================================
def max_profit_optimal(prices: List[int]) -> int:
    """Track running minimum price. At each day, check if selling gives best profit."""
    min_price = prices[0]
    max_profit = 0

    for i in range(1, len(prices)):
        # Could we profit by selling today?
        profit_today = prices[i] - min_price
        max_profit = max(max_profit, profit_today)

        # Update minimum price seen so far
        min_price = min(min_price, prices[i])

    return max_profit


# ============================================================
# DRIVER
# ============================================================
if __name__ == "__main__":
    print("=== Best Time to Buy and Sell Stock I ===\n")

    test_cases = [
        ([7, 1, 5, 3, 6, 4], 5),
        ([7, 6, 4, 3, 1], 0),
        ([2, 4, 1], 2),
        ([1, 2], 1),
        ([5], 0),
        ([3, 3, 3, 3], 0),
        ([1, 2, 3, 4, 5], 4),
    ]

    for prices, expected in test_cases:
        b = max_profit_brute(prices)
        o = max_profit_optimal(prices)
        status = "PASS" if b == expected and o == expected else "FAIL"
        print(f"prices = {prices}")
        print(f"  Brute: {b} | Optimal: {o} | Expected: {expected}  [{status}]\n")

File: step_16_dynamic_programming/16.6_dp_on_stocks/P001_best_time_to_buy_and_sell_i/Solution.java
/**
 * Problem: Best Time to Buy and Sell Stock I (LeetCode #121)
 * Difficulty: EASY | XP: 10
 *
 * Given array prices, find max profit from one buy and one sell.
 *
 * @author DSA_Nova
 */
import java.util.Arrays;

// ============================================================
// Approach 1: Brute Force (All Pairs)
// Time: O(n^2) | Space: O(1)
// ============================================================
class BruteStock {
    public int maxProfit(int[] prices) {
        int maxProfit = 0;
        int n = prices.length;

        for (int i = 0; i < n - 1; i++) {
            for (int j = i + 1; j < n; j++) {
                int profit = prices[j] - prices[i];
                maxProfit = Math.max(maxProfit, profit);
            }
        }

        return maxProfit;
    }
}

// ============================================================
// Approach 2: Optimal (Single Pass with Running Minimum)
// Time: O(n) | Space: O(1)
// ============================================================
class OptimalStock {
    public int maxProfit(int[] prices) {
        int minPrice = prices[0];
        int maxProfit = 0;

        for (int i = 1; i < prices.length; i++) {
            // Could we profit by selling today?
            int profitToday = prices[i] - minPrice;
            maxProfit = Math.max(maxProfit, profitToday);

            // Update minimum price seen so far
            minPrice = Math.min(minPrice, prices[i]);
        }

        return maxProfit;
    }
}

// ============================================================
// Main driver
// ============================================================
public class Solution {
    public static void main(String[] args) {
        System.out.println("=== Best Time to Buy and Sell Stock I ===\n");

        BruteStock brute = new BruteStock();
        OptimalStock optimal = new OptimalStock();

        int[][] testCases = {
            {7, 1, 5, 3, 6, 4},
            {7, 6, 4, 3, 1},
            {2, 4, 1},
            {1, 2},
            {5},
            {3, 3, 3, 3},
            {1, 2, 3, 4, 5},
        };
        int[] expected = {5, 0, 2, 1, 0, 0, 4};

        for (int t = 0; t < testCases.length; t++) {
            int b = brute.maxProfit(testCases[t]);
            int o = optimal.maxProfit(testCases[t]);

            boolean pass = b == expected[t] && o == expected[t];
            System.out.println("prices = " + Arrays.toString(testCases[t]));
            System.out.println("  Brute: " + b + " | Optimal: " + o
                    + " | Expected: " + expected[t] + " | Pass: " + pass);
            System.out.println();
        }
    }
}
